contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -0,0 +1,821 @@
+  if (dlib == nullptr) {
+    ENVOY_LOG(error, "golang extension filter dynamicLib is nullPtr.");
+    return 0;
+  }
+
+  std::string str;
+  if (!plugin_config_.SerializeToString(&str)) {
+    ENVOY_LOG(error, "failed to serialize any pb to string");
+    return 0;
+  }
+  auto ptr = reinterpret_cast<unsigned long long>(str.data());
+  auto len = str.length();
+  config_id_ = dlib->envoyGoFilterNewHttpPluginConfig(ptr, len);
+  if (config_id_ == 0) {
+    ENVOY_LOG(error, "invalid golang plugin config");
+  }
+  return config_id_;
+}
+
+FilterConfigPerRoute::FilterConfigPerRoute(
+    const envoy::extensions::filters::http::golang::v3alpha::ConfigsPerRoute& config,
+    Server::Configuration::ServerFactoryContext&) {
+  // NP: dso may not loaded yet, can not invoke envoyGoFilterNewHttpPluginConfig yet.
+  ENVOY_LOG(debug, "initilizing per route golang filter config");
+
+  for (const auto& it : config.plugins_config()) {
+    auto plugin_name = it.first;
+    auto route_plugin = it.second;
+    RoutePluginConfigPtr conf(new RoutePluginConfig(route_plugin));
+    ENVOY_LOG(debug, "per route golang filter config, type_url: {}",
+              route_plugin.config().type_url());
+    plugins_config_.insert({plugin_name, std::move(conf)});
+  }
+}
+
+uint64_t FilterConfigPerRoute::getPluginConfigId(uint64_t parent_id, std::string plugin_name,
+                                                 std::string so_id) const {
+  auto it = plugins_config_.find(plugin_name);
+  if (it != plugins_config_.end()) {
+    return it->second->getMergedConfigId(parent_id, so_id);
+  }
+  ENVOY_LOG(debug, "golang filter not found plugin config: {}", plugin_name);
+  // not found
+  return parent_id;
+}
+
+uint64_t RoutePluginConfig::getMergedConfigId(uint64_t parent_id, std::string so_id) {
+  if (merged_config_id_ > 0) {
+    return merged_config_id_;
+  }
+  auto dlib = Dso::DsoInstanceManager::getDsoInstanceByID(so_id);
+  if (dlib == nullptr) {
+    ENVOY_LOG(error, "golang extension filter dynamicLib is nullPtr.");
+    return 0;
+  }
+
+  if (config_id_ == 0) {
+    std::string str;
+    if (!plugin_config_.SerializeToString(&str)) {
+      ENVOY_LOG(error, "failed to serialize any pb to string");
+      return 0;
+    }
+    auto ptr = reinterpret_cast<unsigned long long>(str.data());
+    auto len = str.length();
+    config_id_ = dlib->envoyGoFilterNewHttpPluginConfig(ptr, len);
+    if (config_id_ == 0) {
+      ENVOY_LOG(error, "invalid golang plugin config");
+      return parent_id;
+    }
+    ENVOY_LOG(debug, "golang filter new plugin config, id: {}", config_id_);
+  }
+
+  merged_config_id_ = dlib->envoyGoFilterMergeHttpPluginConfig(parent_id, config_id_);
+  if (merged_config_id_ == 0) {
+    ENVOY_LOG(error, "invalid golang plugin config");
+  }
+  ENVOY_LOG(debug, "golang filter merge plugin config, from {} + {} to {}", parent_id, config_id_,
+            merged_config_id_);
+  return merged_config_id_;
+};
```

Comment:
All of the errors in this function should I think be caught at config load time so should be throwing exceptions. I'm not sure what happens in the error path here. Please verify this and add appropriate config loading tests that fail.